torch/csrc/jit/codegen/cuda/lower_warp_reduce.cpp

```diff
@@ -0,0 +1,546 @@
+      } else {
+        if (shouldEliminate(expr)) {
+          result_expr = nullptr;
+        }
+      }
+
+      // Push the result expr if not eliminated
+      if (result_expr) {
+        result_exprs.push_back(result_expr);
+      }
+    }
+
+    return result_exprs;
+  }
+
+  kir::ForLoop* eliminateDeadCode(kir::ForLoop* for_loop) {
+    auto new_loop_body = eliminateDeadCodeInScope(for_loop->body().exprs());
+    if (new_loop_body.empty()) {
+      return nullptr;
+    }
+
+    // TODO: we will need a kernel_ir cloner to make this
+    //  kind of logic re-usable.
+    auto new_loop = ir_builder_.create<kir::ForLoop>(
+        for_loop->iter_domain(),
+        for_loop->index(),
+        for_loop->start(),
+        for_loop->stop(),
+        for_loop->step(),
+        for_loop->vectorize(),
+        for_loop->vectorize_shift());
+    for (auto expr : new_loop_body) {
+      new_loop->body().push_back(expr);
+    }
+    return new_loop;
+  }
+
+  kir::IfThenElse* eliminateDeadCode(kir::IfThenElse* ite) {
+    auto new_then_body = eliminateDeadCodeInScope(ite->thenBody().exprs());
+    auto new_else_body = eliminateDeadCodeInScope(ite->elseBody().exprs());
+    if (new_then_body.empty() && new_else_body.empty()) {
+      return nullptr;
+    }
+    auto new_ite = ir_builder_.create<kir::IfThenElse>(ite->predicate());
+
+    for (auto expr : new_then_body) {
+      new_ite->thenBody().push_back(expr);
+    }
+    for (auto expr : new_else_body) {
+      new_ite->elseBody().push_back(expr);
+    }
+    return new_ite;
+  }
+
+ private:
+  std::unordered_set<kir::TensorView*> live_tvs_;
+  std::unordered_set<kir::TensorView*> dead_tvs_;
+  std::vector<kir::TensorView*> candidate_tvs_;
+  std::unordered_set<kir::TensorView*> candidate_tv_set_;
+
+  std::vector<kir::Expr*> result_exprs_;
+  kir::IrBuilder ir_builder_;
+};
+
+//! A pass to eliminate redundant parallel broadcasts that are consumers
+//!  of warp reduction.
+//! Detects the following pattern:
+//!
+//!  For ... (serial)
+//!   For ... (serial)
+//!     T1[0] = warp_reduce (T0[0])
+//!     T2[0] = block_broadcast (T1[0])
+//!
+//!  The block_broadcast can then be eliminated given that both the warp
+//!   reduce and the broadcast are known in compile-time to be parallelized
+//!   on a single warp only.
+//!
+//!  Currently only limited to buffers of size-1 to avoid having to
+//!   re-run indexing
+//!
+//!  This pass operates in 3 phases:
+//!   1. fuseBroadcastWithWarpReduce identifies the broadcasts that can
+//!    be removed, and generates a replacement map from the broadcast
+//!    output to reduction output.
+//!
+//!   2. kir_utils::replaceInputsInExpr replaces applicable uses of
+//!    the broadcast output with the corresponding reduction output.
+//!
+//!   3. eliminateDeadBroadcastAndAllocate removes the broadcast ops
+//!    and corresponding allocations if they're un-used after step 2.
+class fuseBroadcastWithWarpReduce {
```

Comment:
nit: class name should start with upper case.

Reply:
Updated thanks!